src/mystery_gift.c

```diff
@@ -0,0 +1,293 @@
+
+void Save_MysteryGift_Init(MYSTERY_GIFT_SAVE* mg) {
+    SaveSubstruct_UpdateCRC(SAVE_MYSTERY_GIFT);
+}
+
+BOOL MysteryGiftTagIsValid(u32 tag) {
+    return tag != MG_TAG_invalid && tag < MG_TAG_max;
+}
+
+MysteryGift* SaveMysteryGift_GetByIdx(MYSTERY_GIFT_SAVE* mg, int index) {
+    if (index >= 0 && index < NUM_SAVED_MYSTERY_GIFTS) {
+        MysteryGift* ret = &mg->gifts[index];
+        if (MysteryGiftTagIsValid(ret->tag)) {
+            return ret;
+        }
+    }
+
+    return NULL;
+}
+
+WonderCard* SaveMysteryGift_CardGetByIdx(MYSTERY_GIFT_SAVE* mg, int index) {
+    if (index >= 0 && index < NUM_SAVED_WONDER_CARDS) {
+        WonderCard* ret = &mg->cards[index];
+        if (MysteryGiftTagIsValid(ret->tag)) {
+            return ret;
+        }
+    } else if (index == RECEIVED_WONDER_CARD_IDX) {
+        return &mg->specialWonderCard;
+    }
+
+    return NULL;
+}
+
+BOOL SaveMysteryGift_TryInsertGift(MYSTERY_GIFT_SAVE* mg, const MysteryGift* src, int flag) {
+    BOOL ret = FALSE;
+    int i;
+    
+    if (!SaveMysteryGift_FindAvailable(mg)) {
+        return FALSE;
+    }
+
+    for (i = 0; i < NUM_SAVED_MYSTERY_GIFTS; ++i) {
+        if (!MysteryGiftTagIsValid(mg->gifts[i].tag)) {
+            MI_CpuCopy8(src, &mg->gifts[i], sizeof(MysteryGift));
+            mg->gifts[i].flag = flag;
+            ret = TRUE;
+            break;
+        }
+    }
+
+    SaveSubstruct_UpdateCRC(SAVE_MYSTERY_GIFT);
+    return ret;
+}
+
+BOOL SaveMysteryGift_TryInsertCard(MYSTERY_GIFT_SAVE* mg, const WonderCard* src) {
+    BOOL ret = FALSE;
+    int i;
+
+    if (!SaveMysteryGift_CardFindAvailable(mg)) {
+        return FALSE;
+    }
+
+    if (src->unk104.unk8E_3 == 1 && !SaveMysteryGift_FindAvailable(mg)) {
+        return FALSE;
+    }
+
+    for (i = 0; i < NUM_SAVED_WONDER_CARDS; ++i) {
+        if (!MysteryGiftTagIsValid(mg->cards[i].tag)) {
+            MI_CpuCopy8(src, &mg->cards[i], sizeof(WonderCard));
+            if (src->unk104.unk8E_3 == 1) {
+                SaveMysteryGift_TryInsertGift(mg, (const MysteryGift*)src, i);
+            }
+            ret = TRUE;
+            break;
+        }
+    }
+
+    SaveSubstruct_UpdateCRC(SAVE_MYSTERY_GIFT);
+    return ret;
+}
+
+BOOL SaveMysteryGift_TrySetSpecialCard(MYSTERY_GIFT_SAVE* mg, const WonderCard* src) {
+    if (MysteryGiftTagIsValid(mg->specialWonderCard.tag) == TRUE) {
+        return FALSE;
+    }
+
+    MI_CpuCopy8(src, &mg->specialWonderCard, sizeof(WonderCard));
+    SaveSubstruct_UpdateCRC(SAVE_MYSTERY_GIFT);
+    return FALSE;
+}
+
+BOOL SaveMysteryGiftI_SetReceived(MYSTERY_GIFT_SAVE* mg, int index) {
+    GF_ASSERT(index < NUM_SAVED_MYSTERY_GIFTS);
+    mg->gifts[index].tag = MG_TAG_invalid;
+    mg->gifts[index].flag = 0;
+    SaveSubstruct_UpdateCRC(SAVE_MYSTERY_GIFT);
+    return TRUE;
+}
+
+BOOL SaveMysteryGift_ReceiveGiftAndClearCardByIndex(MYSTERY_GIFT_SAVE* mg, int index) {
+    GF_ASSERT(index < NUM_SAVED_WONDER_CARDS);
+    mg->cards[index].tag = MG_TAG_invalid;
```

Comment:
MG_TAG_INVALID?

Reply:
Would need to change all the other constants in this define list.